design/design-doc-e2e-testing.md

```diff
@@ -0,0 +1,861 @@
+maintainable in the long run.
+
+### E2E-framework
+
+[kubernetes-sig/e2e-framework] is a new framework for writing imperative
+end-to-end tests for Kubernetes.
+
+#### Anatomy of a test
+
+Tests are organized in test suites, each test suite being a Go package. The
+function TestMain is used to define package-wide testing steps and configure 
+behavior. The framework exposes the Environment interface, which is used to
+interact with the cluster.
+
+Furthermore, the environment is set up by defining steps that are executed in
+different phases of the test, like Setup, Finish, BeforeEach, kinda similar to
+Ginkgo.
+
+TestMain:
+
+```go
+var testenv *env.Environment
+
+func TestMain(m *testing.M) {
+	testenv = env.NewInClusterConfig()
+
+	testenv.Setup(
+		setupSchema,
+		setupDummyProvider,
+	)
+
+	testenv.Finish(
+		teardownDummyProvider,
+	)
+
+	// launch package tests
+	os.Exit(testenv.Run(m))
+}
+```
+
+The tests themselves are organized in Features for which steps (Setup, Assess,
+Teardown) need to be defined.
+
+Test example:
+
+```go
+func TestFlow(t *testing.T) {
+
+	f := features.New("create claim flow").
+		WithLabel("type", "flow-claim").
+		WithSetup("install the XRD", setupXRD).
+		WithTeardown("teardown the XRD", teardownXRD).
+		WithSetup("install the Composition", setupComposition).
+		WithTeardown("teardown the Composition", teardownComposition).
+		WithSetup("xrd is established and offered", func(ctx context.Context, t *testing.T, cfg *envconf.Config) context.Context {
+			xrd := genXRD()
+			err := wait.For(conditions.New(cfg.Client().Resources()).ResourceMatch(xrd, func(object k8s.Object) bool {
+				o := object.(*extv1.CompositeResourceDefinition)
+				return o.Status.GetCondition(extv1.TypeEstablished).Status == corev1.ConditionTrue &&
+					o.Status.GetCondition(extv1.TypeOffered).Status == corev1.ConditionTrue
+			}), wait.WithTimeout(time.Minute*1))
+			if err != nil {
+				t.Fatalf("failed to wait for XRD to be established and offered: %v", err)
+			}
+			return ctx
+		}).
+		Setup(setupClaim).
+		Teardown(teardownClaim).
+		Assess("claim", func(ctx context.Context, t *testing.T, cfg *envconf.Config) context.Context {
+			claim, err := getClaim()
+			if err != nil {
+				t.Fatalf("failed to get claim: %v", err)
+			}
+			err = wait.For(conditions.New(cfg.Client().Resources()).ResourceMatch(claim, func(object k8s.Object) bool {
+				claimObj := composed.Unstructured{Unstructured: *claim}
+				isReady := claimObj.GetCondition(xpv1.TypeReady)
+				if isReady.Status != corev1.ConditionTrue {
+					t.Logf("claim %q is not yet Ready", claim.GetName())
+					return false
+				}
+				return isReady.Status == corev1.ConditionTrue
+			}), wait.WithTimeout(time.Minute*1))
+			if err != nil {
+				t.Fatalf("failed to wait for claim to be Ready: %v", err)
+			}
+			return ctx
+		}).Feature()
+
+	// test feature
+	testenv.Test(t, f)
+}
+...
+```
+
+The test steps are run in the order based on their level (Setup, Assess,
+Teardown).
```

Comment:
I've mentioned above that Ginkgo handles cleanup also when a test times out or gets interrupted when running it interactively. I'd be happy to be proven wrong, but because e2e-framework is run through `go test`, I'm pretty sure that the cleanup code will not run when `go test` decides that it is time to stop (global timeout, CTRL-C).

When testing against a throwaway cluster that's not an issue (just recreate it), but otherwise leaving garbage behind can be annoying.